pkg/virtual/syncer/strategies/helpers.go

```diff
@@ -0,0 +1,120 @@
+	labels := kcpResource.GetLabels()
+	syncing := make(Syncing, len(labels))
+	annotations := kcpResource.GetAnnotations()
+	for labelName, labelValue := range labels {
+		if strings.HasPrefix(labelName, v1alpha1.InternalClusterResourceStateLabelPrefix) {
+			syncTarget := strings.TrimPrefix(labelName, v1alpha1.InternalClusterResourceStateLabelPrefix)
+			syncTargetSyncing := SyncTargetSyncing{
+				RequestedState: RequestedSyncingState(labelValue),
+				Detail:         annotations[SyncingTransformationAnnotationPrefix+syncTarget],
+			}
+			if deletionAnnotation, exists := annotations[v1alpha1.InternalClusterDeletionTimestampAnnotationPrefix+syncTarget]; exists {
+				var deletionTimestamp metav1.Time
+				if err := deletionTimestamp.UnmarshalText([]byte(deletionAnnotation)); err != nil {
+					klog.Errorf("Parsing of the deletion annotation for sync target %q failed: %v", syncTarget, err)
+				} else {
+					syncTargetSyncing.DeletionTimestamp = &deletionTimestamp
+				}
+			}
+			if finalizersAnnotation, exists := annotations[v1alpha1.ClusterFinalizerAnnotationPrefix+syncTarget]; exists {
+				syncTargetSyncing.Finalizers = finalizersAnnotation
+			}
+			syncing[syncTarget] = syncTargetSyncing
+		}
+	}
+	return syncing
+}
+
+func CarryOnSyncerViewStatus(existingSyncerViewResource, newSyncerViewResource *unstructured.Unstructured) error {
+	if existingSyncerViewResource != nil {
+		// Set the status to the syncer view (last set status for this sync target)
+		if status, exists, err := unstructured.NestedFieldCopy(existingSyncerViewResource.UnstructuredContent(), "status"); err != nil {
+			return err
+		} else if exists {
+			if err := unstructured.SetNestedField(newSyncerViewResource.UnstructuredContent(), status, "status"); err != nil {
+				return err
+			}
+		} else {
+			unstructured.RemoveNestedField(newSyncerViewResource.UnstructuredContent(), "status")
+		}
+	}
+	return nil
+}
+
+func GetSyncerViews(kcpResource *unstructured.Unstructured, syncTargets ...string) (map[string]unstructured.Unstructured, error) {
+	kcpResource = kcpResource.DeepCopy()
+
+	annotations := kcpResource.GetAnnotations()
+
+	filterSyncTargets := len(syncTargets) > 0
+	syncTargetsToKeep := sets.NewString(syncTargets...)
+
+	syncerViewDiffs := make(map[string]string)
+	for name, value := range annotations {
+		if strings.HasPrefix(name, syncerViewAnnotationPrefix) {
+			syncTarget := strings.TrimPrefix(name, syncerViewAnnotationPrefix)
+			if filterSyncTargets && !syncTargetsToKeep.Has(syncTarget) {
+				continue
+			}
+			syncerViewDiffs[syncTarget] = value
+			delete(annotations, name)
+		}
+	}
+	kcpResource.SetAnnotations(annotations)
+	kcpResourceJson, err := kcpResource.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+
+	syncerViewResources := make(map[string]unstructured.Unstructured)
+	for syncTarget, syncerViewDiff := range syncerViewDiffs {
+		syncerViewResourceJson, err := jsonpatch.MergePatch([]byte(kcpResourceJson), []byte(syncerViewDiff))
```

Comment:
have seen this also in kube. Is there really no way to apply merge patches without roundtripping through JSON? This looks like a giant waste of resources.